VoxelSniperCore/src/main/java/com/github/kevindagame/brush/polymorphic/PolyBrush.kt

```diff
@@ -0,0 +1,212 @@
+                    addOperation(BlockOperation(position, position.block.blockData, material.createBlockData()))
+                }
+            }
+        } else {
+            addOperations(currentPerformer.perform(positions, v))
+        }
+    }
+
+    private fun getPositions(v: SnipeData): MutableList<BaseLocation> {
+        val positions = initPositions(v)
+        val newPositions = mutableListOf<BaseLocation>()
+        val radiusSquared = (v.brushSize + if (smooth) SMOOTH_CIRCLE_VALUE else VOXEL_CIRCLE_VALUE).pow(2)
+        val center = targetBlock
+
+        for (position in positions) {
+            for (shape in shapes) {
+                if (shape.apply(v, position, radiusSquared)) {
+                    newPositions.add(
+                        BaseLocation(
+                            world,
+                            (center.x + position.dx).toDouble(),
+                            (center.y + position.dy).toDouble(),
+                            (center.z + position.dz).toDouble()
+                        )
+                    )
+                }
+            }
+        }
+        return newPositions
+    }
+
+    /**
+     * Initialize the positions for every block in the brush radius in the shape of a cube
+     * "A relative location a day keeps the Math pain away" - KevinDaGame
+     */
+    private fun initPositions(v: SnipeData): MutableList<PolyLocation> {
+        val positions = mutableListOf<PolyLocation>()
+        val brushSize = v.brushSize
+        for (z in brushSize downTo -brushSize) {
+            for (x in brushSize downTo -brushSize) {
+                for (y in brushSize downTo -brushSize) {
+                    positions.add(
+                        PolyLocation(x, y, z)
+                    )
+                }
+            }
+        }
+        return positions
+    }
+
+    private val smooth: Boolean
+        get() {
+            for (property in properties) {
+                if (property is SmoothProperty) {
+                    return property.value
+                }
+            }
+            return false
+        }
+    private val excludeAir: Boolean
+        get() {
+            for (property in properties) {
+                if (property is ExcludeAirProperty) {
+                    return property.value
+                }
+            }
+            return false
+        }
+    private val excludeWater: Boolean
+        get() {
+            for (property in properties) {
+                if (property is ExcludeWaterProperty) {
+                    return property.value
+                }
+            }
+            return false
+        }
+    private val currentPerformer: BasePerformer
+        get() {
+            for (property in properties) {
+                if (property is PerformerProperty) {
+                    return property.value
+                }
+            }
+            return PerformerProperty().value
+        }
```

Comment:
I wonder if we could make this generic, since it is four times the same code...

Reply:
https://github.com/KevinDaGame/VoxelSniper-Reimagined/commit/bb01fc97ded16adea1753b9930036ab85adb8e14